src/log.c

```diff
@@ -36,70 +36,378 @@
 #include <stdarg.h>
 #include <stdio.h>
 #include <stdlib.h>
+#include <string.h>
+#include <inttypes.h>
 
+#include <rdma/fi_errno.h>
 #include "fi.h"
 #include "fi_log.h"
 
+#define ARRAY_SIZE(x) (sizeof((x)) / sizeof((x)[0]))
+
+static struct slist provider_handles;
+static struct log_providers provider_list;
+
+const char *subsystems[] = {
+	[FABRIC] = "fabric",
+	[DOMAIN] = "domain",
+	[EP_CM] = "ep_cm",
+	[EP_DM] = "ep_dm",
+	[AV] = "av",
+	[CQ] = "cq",
+	[EQ] = "eq",
+	[MR] = "mr"
+};
+
+const char *log_levels[] = {
+	[LOG_WARN] = "warn",
+	[LOG_TRACE] = "trace",
+	[LOG_INFO] = "info"
+};
+
+/*
+ * By default support logging from all providers and subsystems. This will
+ * change if the relevant environment variables are set.
+ */
+uint64_t log_mask = PROV_MASK | SUBSYS_MASK;
+
+typedef void found_func_t(char *ptr);
+
 /* General implementation note: these functions currently use multiple fprintfs
  * in a row, which can render in an ugly fashion for multithreaded code and for
  * some mpirun implementations.  If this bugs anyone enough then we can convert
  * them to snprintf to build up the printout in a single buffer.
  */
 
-int fi_log_level = INT_MIN;
+static int compare_names(struct slist_entry *item, const void *arg)
+{
+	struct provider_parameter *target =
+	    container_of(item, struct provider_parameter, entry);
+	return !strcmp(target->name, arg);
```

Comment:
Yes, strcasecmp.  Other platforms may need to implement their own, if one is not provided.